api/capi/include/nntrainer.h

```diff
@@ -65,9 +72,31 @@ typedef enum {
   ML_LAYER_TYPE_UNKNOWN    /**< Unknown Lyaer */
 } ml_layer_type_e;
 
+/**
+ * @brief Dataset generator callback function for train/valid/test data.
+ * @details The Containers passed will already be allocated with sufficient
+ * space to contain the data by the caller. This function should return a batch
+ * of data in the passed containers.
+ * @note This function can be called multiple times in parallel.
+ * @param[out] input Container to provide input data to the model.
+ * @param[out] label Container to provide label data to the model to train/test.
+ * @param[out] last Container to provide is the data is finished.
+ * @return @c 0 on success. Otherwise a negative error value.
+ * @retval #ML_ERROR_NONE Successful.
+ * @retval #ML_ERROR_INVALID_PARAMETER Invalid Parameter.
+ */
+typedef int (*dataset_generator) (float *input, float *label, bool *last);
+
+/**
+ * @brief Callback function to notify completion of training of the model..
+ * @param[in] cb The callback to be called once the training is finished.
+ * @param[in] data Internal data to be given to the callback, cb.
+ */
+typedef void (*train_notify) (ml_nnmodel_h model, void *data);
```

Comment:
AFAIK, train runs synchronously for now. Could you check this out?

Reply:
Yes, it will still be synchronous. Their will be wrapper around it to make it asynchronous. This for `ml_nnmodel_train_async`

Reply:
Sorry for that. I didn't read that part yet. Thanks for the quick response :)